idx/bigtable/bigtable.go

```diff
@@ -259,56 +259,63 @@ func (b *BigtableIdx) AddOrUpdate(mkey schema.MKey, data *schema.MetricData, par
 		// So we need to explicitly delete the old entry.
 		if oldPartition != partition {
 			go func() {
-				err := b.deleteRow(FormatRowKey(archive.Id, oldPartition))
+				err := b.deleteRow(FormatRowKey(id, oldPartition))
 				if err != nil {
-					log.Errorf("bigtable-idx: Failed to delete row %s: %s", archive.Id, err)
+					log.Errorf("bigtable-idx: Failed to delete row %s: %s", id, err)
 				}
 			}()
 		}
 	}
 
 	// check if we need to save to bigtable.
 	now := uint32(time.Now().Unix())
-	if archive.LastSave < (now - b.cfg.updateInterval32) {
-		archive = b.updateBigtable(now, inMemory, archive, partition)
+	if lastSave < (now - b.cfg.updateInterval32) {
+		b.updateBigtable(now, inMemory, id, partition, lastSave)
 	}
 
 	stat.Value(time.Since(pre))
-	return archive, oldPartition, inMemory
+	return id, lastSave, oldPartition, inMemory
 }
 
 // updateBigtable saves the archive to bigtable and
 // updates the memory index with the updated fields.
-func (b *BigtableIdx) updateBigtable(now uint32, inMemory bool, archive *interning.ArchiveInterned, partition int32) *interning.ArchiveInterned {
+func (b *BigtableIdx) updateBigtable(now uint32, inMemory bool, id schema.MKey, partition int32, lastSave uint32) {
 	// if the entry has not been saved for 1.5x updateInterval
 	// then perform a blocking save.
-	if archive.LastSave < (now - b.cfg.updateInterval32 - (b.cfg.updateInterval32 / 2)) {
+	if lastSave < (now - b.cfg.updateInterval32 - (b.cfg.updateInterval32 / 2)) {
 		if log.IsLevelEnabled(log.DebugLevel) {
-			log.Debugf("bigtable-idx: updating def %s in index.", archive.Id)
+			log.Debugf("bigtable-idx: updating def %s in index.", id)
+		}
+		arc, ok := b.MemoryIndex.Get(id)
+		if ok {
+			b.writeQueue <- writeReq{recvTime: time.Now(), def: arc.CloneInterned()}
```

Comment:
this .CloneInterned() call shouldn't be necessary as long as arc isn't read after it's been pushed into the write queue

Reply:
arc is a `*interning.ArchiveInterned`, so we still need to `Clone()` it and release it after the Queue is done processing it. Same for Bigtable.

Reply:
yes, Clone / Release are to inc / dec reference count. so when you are sure that after pushing that element into the queue it doesn't get accessed in this routine anymore you can just consider the reference to "travel" through the queue, and it will then need to be released by the consumer of the queue